daemon/logger/journald_semistruct/journald_semistruct.go

```
@@ -0,0 +1,140 @@
+	if err := logger.RegisterLogDriver(name, New); err != nil {
+		logrus.Fatal(err)
+	}
+	if err := logger.RegisterLogOptValidator(name, validateLogOpt); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+// New creates a journald logger using the configuration passed in on
+// the context.
+func New(ctx logger.Context) (logger.Logger, error) {
+	if !journal.Enabled() {
+		return nil, fmt.Errorf("journald is not enabled on this host")
+	}
+	// Strip a leading slash so that people can search for
+	// CONTAINER_NAME=foo rather than CONTAINER_NAME=/foo.
+	name := ctx.ContainerName
+	if name[0] == '/' {
+		name = name[1:]
+	}
+
+	// parse log tag
+	tag, err := loggerutils.ParseLogTag(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+
+	vars := map[string]string{
+		"CONTAINER_ID":      ctx.ContainerID[:12],
+		"CONTAINER_ID_FULL": ctx.ContainerID,
+		"CONTAINER_NAME":    name,
+		"CONTAINER_TAG":     tag,
+	}
+
+	extraAttrs := ctx.ExtraAttributes(strings.ToTitle)
+
+	for k, v := range extraAttrs {
+		vars[k] = v
+	}
+
+	pr := sp.ParseSemistruct()
+	return &journald{vars: vars, parser: pr, readers: readerList{readers: make(map[*logger.LogWatcher]*logger.LogWatcher)}}, nil
+}
+
+// We don't actually accept any options, but we have to supply a callback for
+// the factory to pass the (probably empty) configuration map to.
+func validateLogOpt(cfg map[string]string) error {
+	for key := range cfg {
+		switch key {
+		case "labels":
+		case "env":
+		case "tag":
+		default:
+			return fmt.Errorf("unknown log opt '%s' for journald log driver", key)
+		}
+	}
+	return nil
+}
+
+func (s *journald) Log(msg *logger.Message) error {
+
+	var semistruct_line cp.Match
+
+	journald_vars := s.vars
+
+	line := string(msg.Line)
+
+	// Peak at the first few characters, if they start with the
+	// sentinel then attempt a parse, otherwise don't parse and just
+	// shove the whole line out to journald.
+	if len(line) > 2 && line[:2] == "!<" {
+		semistruct_line, _ = s.parser.ParseString(line)
```

camelcase, why are you ignoring error?